Add Skygear Chat UIKit Android & iOS Guide

howawong opened the pull request:
connects #165

howawong committed:
Add Skygear Chat UIKit Android & iOS Guide

refs #165

carmenlau commented:
In both iOS and Android installation instructions should mention the required permission setup

howawong commented:
@carmenlau, in Android, `chat` and `chat_ui` `AndroidManifest.xml` include the permission already.

howawong commented:
@carmenlau, I checked with @ben181231, it does not take extra permission except push notification.

carmenlau commented on chat/uikit/android.md:
````diff
@@ -0,0 +1,307 @@
+
+```java
+Intent intent = new Intent(this, ConversationActivity.class);
+intent.putExtra(ConversationActivity.ConversationIntentKey, c.toJson().toString()); // c is an instance of an io.skygear.plugins.chat.Conversation
````
Please put the comment to another line, it is difficult to read :P

carmenlau commented on chat/uikit/android.md:
````diff
@@ -0,0 +1,307 @@
+
+```kotlin
+val intent = Intent(this, ConversationActivity::class.java)
+intent.putExtra(ConversationActivity.ConversationIntentKey, c.toJson().toString())
````
Please add the comment to the kotlin version too

howawong replied:
updated.

carmenlau commented on chat/uikit/android.md:
````diff
@@ -0,0 +1,307 @@
+UIKit provides `ConversationViewAdapter` to enable remote HTTP background image in `ConversationActivity`.
+
+```java
+ConversationViewAdapter adapter = new ConversationViewAdapter() {
````
Missing example for how to set `ConversationViewAdapter` to Conversation Activity?

howawong replied:
updated.

carmenlau commented on chat/uikit/ios.md:
````diff
@@ -0,0 +1,386 @@
+- To enable or disable date
+
+```obj-c
+(bool) conversationViewController:(SKYChatConversationViewController *)controller shouldShowDateAt:(NSIndexPath *)indexPath
````
Missing `-` before `(bool)`

carmenlau commented on chat/uikit/ios.md:
````diff
@@ -0,0 +1,386 @@
+
+- To customize date format
+```obj-c
+- (NSAttributedString *) conversationViewController:(SKYChatConversationViewController *)controller dateStringAt:(NSIndexPath *)indexPath
````
Please put implementation example for those delegate functions which are not just return bool. That would be easier to understand :)

howawong replied:
updated.

carmenlau commented on chat/uikit/ios.md:
````diff
@@ -0,0 +1,386 @@
+```obj-c
+[[[SKYChatUIModelCustomization default]
+    updateWithUserNameField:@"name"]
+    updateWithUserAvatarField:@"profile_pic"];
````
I think it is not common to chain the call this way in objective-c? Would be better to call the update separately like this?
```
[[SKYChatUIModelCustomization default] updateWithUserNameField:@"name"];
[[SKYChatUIModelCustomization default] updateWithUserAvatarField:@"profile_pic"];
```

carmenlau commented on chat/uikit/ios.md:
````diff
@@ -0,0 +1,386 @@
+1. add `pod 'SKYKitChat/UI', '~> 1.1'` in `Podfile`.
+
+```bash
+pod 'SKYKitChat/UI', '~> 1.1'
````
Just afraid we will forget to update this version, we should be a newer version that include the ui kit update. Should be 1.2 or 1.4? We can discuss.

howawong replied:
@carmenlau  what is the latest stable release? Is it 1.4 ? If yes, then we just put 1.4.

howawong replied:
@rickmak  any thought on versioning?

carmenlau reviewed (changes requested):
The content is good:)! But the section organization makes me a bit difficult to find the information i want.

Currently ios has following sections:
```
About UIKit
Prerequisite
Creating a conversation UI
Customizations
UICustomization
SKYChatUIModelCustomization
Delegates
```
And android:
```
About UIKit
Prerequisite
Creating a conversation UI
Customizations
Attribute Customization
Event Hooks
Adapter
Avatar Adapter
```

For example, ui customization and event hooks are quite different area. Maybe we should separate them clearly? 
Feel free to discuss how to make better organization for the sections.

howawong committed:
Update Chat UIKit guide based on @carmenlau comments

Connects #165

howawong committed:
Update SKYKitChat to 1.4

howawong commented:
@carmenlau  updated version in `Podfile` to 1.4

carmenlau commented on chat/uikit/ios.md:
```diff
@@ -0,0 +1,408 @@
+|`avatarTextColor`|Text color of the avatar|Color|
+|`userAvatarType`|Avatar Type, either showing initial of the username field or image|`SKYChatConversationViewUserAvatarType`, "initial" or "image"|
+|`avatarHiddenForIncomingMessages`|Hide incoming message avatar if true|Boolean, default False|
+|`avatarHiddenForOutgoingMessages`|Hide outgoing message avatar if true|Boolean, default True|
```
Check the code the `avatarHiddenForOutgoingMessages` default is False

carmenlau commented on chat/uikit/ios.md:
```diff
@@ -0,0 +1,408 @@
+|`messageStatusSomeRead`|Some read text|String, default "Some Read"|
+|`messageStatusAllRead`|All read text|String, default "All Read"|
+|`messageSentFailed`|Failed text|String, default "Failed"|
+|`messageDateFormatter`|Message delivery time date formatter|`DateFormatter`|
```
Text customization is under `textCustomization` in  not `UICustomization`

carmenlau replied:
Missing send button

carmenlau replied:
`messageDateFormatter` is not in this section, it is under UICustomization

carmenlau commented on chat/uikit/ios.md:
```diff
@@ -0,0 +1,408 @@
+|---------|-----------|------|
+|`avatarBackgroundColor`|Background color of the avatar|Color|
+|`avatarTextColor`|Text color of the avatar|Color|
+|`userAvatarType`|Avatar Type, either showing initial of the username field or image|`SKYChatConversationViewUserAvatarType`, "initial" or "image"|
```
No `userAvatarType`, it should be `avatarType`.

Please also add how to select the avatar field in this case

carmenlau replied:
It's still showing user initial with following config
 
```
 [[SKYChatUIModelCustomization default] updateWithUserAvatarField:@"profile_pic"];
 [SKYChatConversationView UICustomization].avatarType = SKYChatConversationViewUserAvatarTypeImage;
```

carmenlau commented on chat/uikit/ios.md:
````diff
@@ -0,0 +1,408 @@
+```obj-c
+SKYChatConversationViewController* conversationViewController = [[SKYChatConversationViewController alloc] init];
+conversationViewController.conversation = c;
+conversationViewController.incomingMessageBubbleColor = [UIColor redColor];
````
We are not using this way to customize `incomingMessageBubbleColor` this way. We used customization class and delegate

carmenlau commented on chat/uikit/ios.md:
```diff
@@ -0,0 +1,408 @@
+|`shouldShowVoiceMessageButton`|Show voice message button if true|Bool, default: True|
+|`shouldShowTypingIndicator`|Show typing indicators if true|Bool, default: True|
+|`shouldShowMessageStatus`|Show message status if true|Bool, default: true|
+|`shouldShowAccessoryButton`|Show accessory button if true|Bool, default: true|
```
Those name are not correct, from the code they are `cameraButtonShouldShow`

carmenlau replied:
We don't have accessoryButton setting in UICustomization

carmenlau replied:
I think we should add `accessoryButtonShouldShow` for customization class to keep the api consistent. User we use both customization class and delegate for this setting. Let me open issue in iOS that this. Please keep it in the document.

carmenlau commented on chat/uikit/ios.md:
```diff
@@ -0,0 +1,408 @@
+|---------|-----------|------|
+|`shouldShowCameraButton`|Show camera button if true|Bool, default: True|
+|`shouldShowVoiceMessageButton`|Show voice message button if true|Bool, default: True|
+|`shouldShowTypingIndicator`|Show typing indicators if true|Bool, default: True|
```
Please use `true` instead of `True`

carmenlau commented on chat/uikit/ios.md:
```diff
@@ -0,0 +1,408 @@
+    NSString *dateString = [formatter stringFromDate: [msg creationDate]];
+    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:dateString];
+    [attributedString setTextColor:[UIColor blueColor]];
+    return attributedString;
```
```
    SKYMessage* msg = [[controller messageList] messageAt: indexPath.row];
    NSDateFormatter * formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat: @"yyyy-MM-dd"];
    NSString *dateString = [formatter stringFromDate: [msg creationDate]];

    NSDictionary *attrs = @{ NSForegroundColorAttributeName : [UIColor blueColor] };
    NSAttributedString *attrStr = [[NSAttributedString alloc] initWithString:dateString attributes:attrs];
    return attrStr;
```

carmenlau commented:
We also missed the code to teach user how to set the conversation view background 🙈

carmenlau commented:
The background image code for your reference:

Customization class
```
    [SKYChatConversationView UICustomization].backgroundColor = [UIColor lightGrayColor];
    [SKYChatConversationView UICustomization].backgroundImage = [UIImage imageNamed:<imageName>];
    [SKYChatConversationView UICustomization].backgroundImageURL = [NSURL URLWithString:<imageURL>];
```

Delegate
```
- (UIColor*) backgroundColorForConversationViewController:(SKYChatConversationViewController *)controller {
    return [UIColor lightGrayColor];
}

- (UIImage*) backgroundImageForConversationViewController:(SKYChatConversationViewController *)controller {
    return [UIImage imageNamed:<imageName>];
}

- (NSURL*) backgroundImageURLForConversationViewController:(SKYChatConversationViewController *)controller {
    return [NSURL URLWithString:<imageURL>];
}
```

carmenlau commented:
@howawong I found we only includes some of the event hooks here, we have more events hooks actually. Also do you think it would be better to open another page for UIKit event hooks? If you agree, let's remove the event hooks from this pr and we will focuses on the ui customization in this PR first. And another pr and issue for the event hooks?

btw, the participants hooks is not updated, the function should be `didFetchParticipants` and receive list of `SKYParticipant` instead of records

https://github.com/SkygearIO/chat-SDK-iOS/blob/02aa80cca37b61708bb7625640021f98e9f4a55f/SKYKitChat/Classes/UI/ConversationView/SKYChatConversationViewController.swift#L123

howawong committed:
Update iOS guide based on @carmenlau comment

- Update objective-C code snippet
- Delete Controller's setter
- Add background related UICustomization and delegate

Connects #165

howawong committed:
Remove event hooks guides in chat UI Kit

Connects #165

carmenlau commented:
@ben181231 Would like to have your help to check the android part

howawong commented:
Merged to `next` branch via `ce2a33f0c3f2fb8abfd4c1df8ee296f4f105b779`. Please create an issue if needed.
